compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -574,21 +499,97 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
 
   /** Can this type be inhabited by a value?
    *
-   *  Intersection between singleton types and other types is always empty
-   *  the singleton type is not a subtype of the other type.
-   *  See patmat/i3573.scala for an example.
+   *  Check is based on the following facts:
+   *
+   *  - single inheritance of classes
+   *  - final class cannot be extended
+   *  - intersection of a singleton type with another irrelevant type  (patmat/i3574.scala)
+   *
    */
-  def inhabited(tpe: Type)(implicit ctx: Context): Boolean = {
-    val emptySingletonIntersection = new ExistsAccumulator({
-      case AndType(s: SingletonType, t) =>
-        !(s <:< t)
-      case AndType(t, s: SingletonType) =>
-        !(s <:< t)
-      case x =>
-        false
-    })
-
-    !emptySingletonIntersection(false, tpe)
+  def inhabited(tp: Type)(implicit ctx: Context): Boolean = {
+    // convert top-level type shape into "conjunctive normal form"
+    def cnf(tp: Type): Type = tp match {
+      case AndType(OrType(l, r), tp)      =>
+        val tp1 = cnf(tp)
+        val l1  = cnf(l)
+        val r1  = cnf(r)
+        OrType(cnf(AndType(l1, tp1)), cnf(AndType(r1, tp1)))
```

Comment:
I think you don't need to compute cnf(l), cnf(r) and cnf(tp) here, it's already handled by the recursion, the body of this case could be changed to the following:

```scala
OrType(cnf(AndType(l, tp)), cnf(AndType(r, tp)))
```

Reply:
You are right, good catch.